### PR TITLE
Require parameters in LectureServlet 

### PR DESCRIPTION
**Pull request by n-wach**
This is a part of #178 

Some additional refactoring was required to bring all the getParameter calls up to the request handler

### CODE REVIEW

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 1]
```diff
@@ -48,11 +45,13 @@
       + "watch\\?v%3D|watch\\?feature=player_embedded&v=|%2Fvideos%2F|"
       + "embed%\u200C\u200B2F|youtu.be%2F|%2Fv%2F)[^#\\&\\?\\n]*";
 
+  @VisibleForTesting static final String PARAM_ID = "id";
```
Comment (amuamushu) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:48]: I left a comment about something similar to this in a different PR.
https://github.com/googleinterns/zoomtube/pull/187#discussion_r469553483

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 2]
```diff
@@ -71,17 +70,41 @@ public void init() throws ServletException {
...
-    Optional<String> optionalVideoUrl = getParameter(request, LINK_INPUT);
-    Optional<Entity> existingEntity = checkUrlInDatabase(optionalVideoUrl);
+    if (request.getParameter(PARAM_NAME) == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Missing name parameter.");
+      return;
+    }
+    String lectureName = request.getParameter(PARAM_NAME);
+
+    if (request.getParameter(PARAM_LINK) == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, "Missing link parameter.");
```
Comment (amuamushu) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:80]: nit: please update the error messages to be constants

File: src/test/java/com/googleinterns/zoomtube/servlets/LectureServletTest.java [hunk 3]
```diff
@@ -83,11 +82,40 @@ public void tearDown() {
     testServices.tearDown();
   }
 
+  @Test
+  public void doPost_missingName_badRequest() throws IOException, ServletException {
+    when(request.getParameter(LectureServlet.PARAM_LINK)).thenReturn(TEST_LINK);
+
+    servlet.doPost(request, response);
+
+    verify(response).sendError(HttpServletResponse.SC_BAD_REQUEST, "Missing name parameter.");
```
Comment (amuamushu) [inline src/test/java/com/googleinterns/zoomtube/servlets/LectureServletTest.java:91]: nit: please make the error message constants

Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/LectureServletTest.java]: Done, thanks!

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 2 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java]: Done, thanks!

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 1 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:48]: Will refactor to make this more logical in #187

Comment (amuamushu) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:48]: sounds good!

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 4]
```diff
@@ -48,11 +45,21 @@
...
+  @VisibleForTesting static final String PARAM_ID = "id";
+
   /* Name of input field used for lecture name in lecture selection page. */
-  private static final String NAME_INPUT = "name-input";
+  @VisibleForTesting static final String PARAM_NAME = "name-input";
 
   /* Name of input field used for lecture video link in lecture selection page. */
-  private static final String LINK_INPUT = "link-input";
+  @VisibleForTesting static final String PARAM_LINK = "link-input";
+
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:55]: nit: you do not need 2 new lines. 1 is fine.

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 5]
```diff
@@ -71,17 +78,41 @@ public void init() throws ServletException {
   // TODO: Check if URL is valid.
   public void doPost(HttpServletRequest request, HttpServletResponse response)
       throws IOException, ServletException {
-    Optional<String> optionalVideoUrl = getParameter(request, LINK_INPUT);
-    Optional<Entity> existingEntity = checkUrlInDatabase(optionalVideoUrl);
+    if (request.getParameter(PARAM_NAME) == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, ERROR_MISSING_NAME);
+      return;
+    }
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:84]: Please do the same here as I mentioned in your other PR.

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 6]
```diff
@@ -97,64 +128,40 @@ private void initializeTranscript(Entity lectureEntity) throws IOException, Serv
...
+  // TODO: Use a filter to avoid fetching all lectures.  See: #185.
+  private Optional<Entity> checkUrlInDatabase(String videoUrl) {
     Query query = new Query(LectureUtil.KIND);
     PreparedQuery results = datastore.prepare(query);
     Iterable<Entity> resultsIterable = results.asIterable();
 
-    String url = videoUrl.get();
     for (Entity lecture : resultsIterable) {
-      if (lecture.getProperty(LectureUtil.VIDEO_URL).equals(url)) {
+      if (lecture.getProperty(LectureUtil.VIDEO_URL).equals(videoUrl)) {
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:158]: Please make this its own PR. it is not related to the PR description.

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 7]
```diff
@@ -97,64 +128,40 @@ private void initializeTranscript(Entity lectureEntity) throws IOException, Serv
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    long lectureId = Long.parseLong(request.getParameter(LectureUtil.ID));
+    if (request.getParameter(PARAM_ID) == null) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, ERROR_MISSING_ID);
+      return;
+    }
+    long lectureId = Long.parseLong(request.getParameter(PARAM_ID));
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:135]: Why are you using "PARAM_ID" instead of LectureUtil.ID?

File: src/test/java/com/googleinterns/zoomtube/servlets/LectureServletTest.java [hunk 8]
```diff
@@ -83,11 +82,40 @@ public void tearDown() {
...
+  }
+
   @Test
   public void doPost_urlAlreadyInDatabase_shouldReturnLecture()
       throws IOException, ServletException {
-    when(request.getParameter(LINK_INPUT)).thenReturn(TEST_LINK);
-    datastoreService.put(LectureUtil.createEntity(/* lectureName= */ "", TEST_LINK, TEST_ID));
+    when(request.getParameter(LectureServlet.PARAM_NAME)).thenReturn(TEST_NAME);
+    when(request.getParameter(LectureServlet.PARAM_LINK)).thenReturn(TEST_LINK);
+    datastoreService.put(LectureUtil.createEntity(TEST_NAME, TEST_LINK, TEST_ID));
```
Comment (billie-cat) [inline src/test/java/com/googleinterns/zoomtube/servlets/LectureServletTest.java:118]: TEST does not add any value to what the variable is. I believe better names would be:

TEST_NAME => LECTURE_NAME
TEST_LINK => URL_LINK
TEST_ID => LECTURE_ID

What do you think?

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 9]
```diff
@@ -48,11 +45,21 @@
       + "watch\\?v%3D|watch\\?feature=player_embedded&v=|%2Fvideos%2F|"
       + "embed%\u200C\u200B2F|youtu.be%2F|%2Fv%2F)[^#\\&\\?\\n]*";
 
+  @VisibleForTesting static final String PARAM_ID = "id";
+
   /* Name of input field used for lecture name in lecture selection page. */
-  private static final String NAME_INPUT = "name-input";
+  @VisibleForTesting static final String PARAM_NAME = "name-input";
```
Comment (f-camp) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:51]: This can be public.

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 10]
```diff
@@ -48,11 +45,21 @@
...
 
   /* Name of input field used for lecture video link in lecture selection page. */
-  private static final String LINK_INPUT = "link-input";
+  @VisibleForTesting static final String PARAM_LINK = "link-input";
```
Comment (f-camp) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:54]: ditto

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 11]
```diff
@@ -48,11 +45,21 @@
...
+
+
+  @VisibleForTesting static final String ERROR_MISSING_NAME = "Missing name parameter.";
+  @VisibleForTesting static final String ERROR_MISSING_LINK = "Missing link parameter.";
+  @VisibleForTesting static final String ERROR_MISSING_ID = "Missing id parameter.";
+  @VisibleForTesting static final String ERROR_INVALID_LINK = "Invalid video link.";
+  @VisibleForTesting static final String ERROR_FAILED_REDIRECT = "Failed to create redirect URL.";
+  @VisibleForTesting static final String ERROR_LECTURE_NOT_FOUND = "Lecture not found in database.";
```
Comment (f-camp) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:62]: Ditto

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 12]
```diff
@@ -71,17 +78,41 @@ public void init() throws ServletException {
   // TODO: Check if URL is valid.
   public void doPost(HttpServletRequest request, HttpServletResponse response)
```
Comment (f-camp) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:79]: Let's have validations in one function as suggested, this servlet looks like it could definitely work well with that.

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 9 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:51]: "If you expose constants in Java, please mark the modifications with @VisibleForTesting"
This is from the doc: https://docs.google.com/document/d/18T9ef-DRqTncVD377GoIQmOGI84cNKkZ7Nn07YURFTw/edit

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 10 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:54]: "If you expose constants in Java, please mark the modifications with @VisibleForTesting"
This is from the doc: https://docs.google.com/document/d/18T9ef-DRqTncVD377GoIQmOGI84cNKkZ7Nn07YURFTw/edit

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 11 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:61]: "If you expose constants in Java, please mark the modifications with @VisibleForTesting"
This is from the doc: https://docs.google.com/document/d/18T9ef-DRqTncVD377GoIQmOGI84cNKkZ7Nn07YURFTw/edit

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 12 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:91]: Done!

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 7 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:137]: We try to use PARAMs defined within the servlet that uses them.  This is because they're defined in each corresponding javascript file.  If we change LectureUtil.ID, it would be hard to know what needs to also change on the javascript side

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 4 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:55]: Removed

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 5 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:97]: Done

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 6 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:136]: Should be gone now

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 13]
```diff
@@ -168,15 +180,15 @@ private Entity getLectureEntityFromRequest(HttpServletRequest request) {
...
    */
-  private Optional<String> buildRedirectUrl(Entity lectureEntity) {
+  private String buildRedirectUrl(Entity lectureEntity) throws ServletException {
     String lectureId = String.valueOf(lectureEntity.getKey().getId());
-
     try {
-      URIBuilder urlBuilder = new URIBuilder(REDIRECT_URL).addParameter(LectureUtil.ID, lectureId);
-      return Optional.of(urlBuilder.build().toString());
-    } catch (URISyntaxException urlBuilderError) {
-      // TODO: Send a response error.
```
Comment (n-wach) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:178]: This may have belonged in a separate PR, but it greatly simplified how the validation worked in doPost, and took care of a TODO which was closely-related to this PR

File: src/test/java/com/googleinterns/zoomtube/servlets/LectureServletTest.java (hunk 8 above)
Comment (n-wach) [reply src/test/java/com/googleinterns/zoomtube/servlets/LectureServletTest.java:121]: This PR did not create those constants, so I'm not going to change them in this PR.  Added tracking issue #216

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 14]
```diff
@@ -71,17 +74,41 @@ public void init() throws ServletException {
   // TODO: Check if URL is valid.
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:74]: Can this be removed yet?

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 15]
```diff
@@ -71,17 +74,41 @@ public void init() throws ServletException {
...
       throws IOException, ServletException {
-    Optional<String> optionalVideoUrl = getParameter(request, LINK_INPUT);
-    Optional<Entity> existingEntity = checkUrlInDatabase(optionalVideoUrl);
+    Optional<String> error = validatePostRequest(request);
+    if (error.isPresent()) {
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST, error.get());
+      return;
+    }
+
+    String lectureName = request.getParameter(PARAM_NAME);
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:83]: This should move to right before it is used. 

See: 4.8.2.2 Declared when needed
Local variables are not habitually declared at the start of their containing block or block-like construct. Instead, local variables are declared close to the point they are first used (within reason), to minimize their scope. Local variable declarations typically have initializers, or are initialized immediately after declaration.

https://google.github.io/styleguide/javaguide.html#s4.8.2-variable-declarations

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java [hunk 16]
```diff
@@ -94,35 +121,44 @@ private void initializeTranscript(Entity lectureEntity) throws IOException, Serv
     String videoId = (String) lectureEntity.getProperty(LectureUtil.VIDEO_ID);
     transcriptParser.parseAndStoreTranscript(videoId, lectureKey);
   }
-
+
```
Comment (billie-cat) [inline src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:124]:
```suggestion

```

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 7 above)
Comment (billie-cat) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:133]: Are you saying LectureUtil.ID is not the same as PARAM_ID?

File: src/test/java/com/googleinterns/zoomtube/servlets/LectureServletTest.java (hunk 8 above)
Comment (billie-cat) [reply src/test/java/com/googleinterns/zoomtube/servlets/LectureServletTest.java:121]: sgtm

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 7 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:133]: They have the same value

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 14 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:74]: IIRC This TODO was to check if an actual youtube video exists with the provided URL

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 7 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:133]: But they have different meanings.  I think LectureUtil.ID was added so we didn't need to define the same PARAM in multiple servlets, but this is a bad idea because we declare them separately in the javascript.  

What we do everywhere else is define the PARAMs that the servlet's requests will be using within the servlet itself.  There may be some duplicate values, but this way if one needs to change, it doesn't break all the other servlets that were using it.

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 15 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:83]: Done

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 14 above)
Comment (billie-cat) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:74]: Can you rephrase this TODO to be more descriptive?

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 7 above)
Comment (billie-cat) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java:132]: Ok sounds good. Thank you :)

File: src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java (hunk 14 above)
Comment (n-wach) [reply src/main/java/com/googleinterns/zoomtube/servlets/LectureServlet.java]: Done, also added tracking issue #224
